### PR TITLE
Enable NodePool controller to apply generated MachineConfigs

### PR DESCRIPTION
**Pull request by dagrayvid**
**What this PR does / why we need it**:
Add functionality to the NodePool controller `getConfig(...)` function to look in the control plane namespace for any ConfigMaps containing MachineConfigs, with the label `hypershift.openshift.io/operator-generated-config: true` and the label `hypershift.openshift.io/nodePool: <nodepool name>`. This is needed in order for the Node Tuning Operator to apply custom tunings which require setting kernel boot parameters. 

For just one simple example, this is required to [reserve hugepages](https://docs.openshift.com/container-platform/4.11/scalability_and_performance/what-huge-pages-do-and-how-they-are-consumed-by-apps.html).

This change has been tested alongside the change in https://github.com/openshift/cluster-node-tuning-operator/pull/456 which adds the ability to NTO to embed the generated MachineConfigs into ConfigMaps.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
[PSAP-742](https://issues.redhat.com/browse/PSAP-742)
See also the enhancement which outlines the plan for this change: https://github.com/openshift/enhancements/pull/1229

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Comment by dagrayvid** [comment]
/hold

For more manual testing by the NTO team.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 1]
```diff
@@ -1626,6 +1638,22 @@ func (r *NodePoolReconciler) enqueueNodePoolsForConfig(obj client.Object) []reco
...
+		// this ConfigMap is referencing
+		for key := range nodePoolList.Items {
+			if obj.GetAnnotations()[nodePoolAnnotation] == nodePoolList.Items[key].Name &&
+				nodePoolList.Items[key].GetNamespace() == obj.GetNamespace() {
+				result = append(result, reconcile.Request{
+					NamespacedName: client.ObjectKeyFromObject(&nodePoolList.Items[key]),
+				})
+				break
+			}
+		}
```
Comment (dagrayvid) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:1654]: @enxebre ideally here we could enqueueParentNodePool for these ConfigMaps, but they are labeled with the NodePool name only, from the NodePool labels on the Nodes. I might be missing a more elegant solution?

File: docs/content/how-to/node-tuning.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Manage node-level tuning with the Node Tuning Operator
 
-If you would like to set some node-level tuning on the nodes in your hosted cluster, you can use the [Node Tuning Operator](https://docs.openshift.com/container-platform/4.11/scalability_and_performance/using-node-tuning-operator.html). In HyperShift, node tuning can be configured by creating ConfigMaps which contain Tuned objects, and referencing these ConfigMaps in your NodePools. Currently Node Tuning is limited to tunables which the TuneD daemon can apply directly like setting `sysctl` values. Tuning that requires setting kernel boot parameters is not yet supported in HyperShift.
+## Creating a simple TuneD profile for setting sysctl settings
+If you would like to set some node-level tuning on the nodes in your hosted cluster, you can use the [Node Tuning Operator](https://docs.openshift.com/container-platform/4.11/scalability_and_performance/using-node-tuning-operator.html). In HyperShift, node tuning can be configured by creating ConfigMaps which contain Tuned objects, and referencing these ConfigMaps in your NodePools.
```
Comment (jmencak) [inline docs/content/how-to/node-tuning.md:4]: I'd recommend changing the link `s/4.11/latest/`.

File: docs/content/how-to/node-tuning.md [hunk 3]
```diff
@@ -95,4 +96,126 @@ If you would like to set some node-level tuning on the nodes in your hosted clus
...
+              summary=Boot time configuration for hugepages
+              include=openshift-node
+              [bootloader]
+              cmdline_openshift_node_hugepages=hugepagesz=2M hugepages=50
+            name: openshift-node-hugepages
+          recommend:
+          - priority: 20
+            profile: openshift-node-hugepages
+    ```
+    > **_NOTE:_**  The `.spec.recommend.match` field is intentionally left blank. In this case this Tuned will be applied to all Nodes in the NodePool where this ConfigMap is referenced. It is advised to group Nodes with the same hardware configuration into the same NodePool. Not following this practice might result in TuneD operands calculating conflicting kernel parameters for two or more nodes sharing the same NodePool.
```
Comment (jmencak) [inline docs/content/how-to/node-tuning.md:134]: This is IMPORTANT, but let's leave it just a NOTE.  This bogged me down as the first "user" of the new code.  An aside, yesterday I was testing with the `.spec.recommend.match` field targetting a single node in the node pool.  It is probably the reason I was [getting this](https://github.com/openshift/cluster-node-tuning-operator/pull/456#pullrequestreview-1097466483) and something that still needs to be addressed on the NTO side.

File: docs/content/how-to/node-tuning.md [hunk 4]
```diff
@@ -95,4 +96,126 @@ If you would like to set some node-level tuning on the nodes in your hosted clus
...
+
+    Create the ConfigMap in the management cluster:
+    ```
+    oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-hugepages.yaml
+    ```
+
+2. Create a new NodePool manifest YAML file, customize the NodePools upgrade type, and reference the previously created ConfigMap in the `spec.tunedConfig` section before creating it.
```
Comment (jmencak) [inline docs/content/how-to/node-tuning.md:141]: The example below clarifies this, but I believe we should clarify by:
`s/before creating it/before creating it in the management cluster/`.

File: docs/content/how-to/node-tuning.md [hunk 5]
```diff
@@ -95,4 +96,126 @@ If you would like to set some node-level tuning on the nodes in your hosted clus
...
+    ```
+    > **_NOTE:_**  Setting `.spec.management.upgradeType` to `InPlace` is recommended to avoid unnecessary Node recreations when applying the new MachineConfigs. With the `Replace` upgrade type, Nodes will be fully deleted and new nodes will replace them when applying the new kernel boot parameters that are calculated by the TuneD operand.
+
+    Create the NodePool in the management cluster:
+    ```
+    oc --kubeconfig="$MGMT_KUBECONFIG" create -f hugepages-nodepool.yaml
+    ```
+
+
+3. After the Nodes become available, the containerized TuneD daemon will calculate the required kernel boot parameters based on the applied TuneD profile. After the Nodes becomes `Ready` and reboot once to apply the generated MachineConfig, you can verify that the Tuned profile is applied and that the kernel boot parameters have been set.
```
Comment (jmencak) [inline docs/content/how-to/node-tuning.md:181]: `s/becomes/become/`

**Review by jmencak — COMMENTED** [review]
Thank your for the PR, David.  The docs look good, I found only minor nits.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 6]
```diff
@@ -1626,6 +1638,22 @@ func (r *NodePoolReconciler) enqueueNodePoolsForConfig(obj client.Object) []reco
...
 	}
 
+	// Check if the ConfigMap is generated by an operator in the control plane namespace
+	// corresponding to this nodepool.
+	if _, ok := obj.GetLabels()[nodePoolOperatorCreatedConfigLabel]; ok {
+		// Go through the list of NodePools to determine which one
+		// this ConfigMap is referencing
+		for key := range nodePoolList.Items {
+			if obj.GetAnnotations()[nodePoolAnnotation] == nodePoolList.Items[key].Name &&
+				nodePoolList.Items[key].GetNamespace() == obj.GetNamespace() {
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:1648]: `nodePoolList.Items[key].GetNamespace() == obj.GetNamespace()`
This will never be true, the generated config live in the control plane namespace. The NodePool live in a different namespace.

The control plane namespace is $hc.Namespace+$hc.Name (which is the same than hcp.Name), you can find the nodePool ns by dropping hcp.Name from the control plane ns, and use it to properly annotate the config .Annotations[nodePoolAnnotation] and then use enqueueParentNodePool.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 7]
```diff
@@ -1303,6 +1304,17 @@ func (r *NodePoolReconciler) getConfig(ctx context.Context,
 		configs = append(configs, *configConfigMap)
 	}
 
+	// Look for operator generated MachineConfigs from the hosted control plane namespace
+	operatorGeneratedConfigs := &corev1.ConfigMapList{}
+	if err := r.List(ctx, operatorGeneratedConfigs, client.MatchingLabels{
+		nodePoolOperatorCreatedConfigLabel: "true",
+		nodePoolAnnotation:                 nodePool.GetName(),
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:1311]: please don't use an "annotation" name for a label. Also that annotation is meant to have the "ns/name" as value.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 8]
```diff
@@ -72,6 +72,7 @@ const (
 
 	nodePoolAnnotationPlatformMachineTemplate = "hypershift.openshift.io/nodePoolPlatformMachineTemplate"
 	nodePoolCoreIgnitionConfigLabel           = "hypershift.openshift.io/core-ignition-config"
+	nodePoolOperatorCreatedConfigLabel        = "hypershift.openshift.io/operator-generated-config"
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:75]: the name of the variable refers to "created" and the value to "generated", can we use a single term?
Also can we please refer to what generates this? e.g `hypershift.openshift.io/nto-config`

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 9]
```diff
@@ -1303,6 +1304,17 @@ func (r *NodePoolReconciler) getConfig(ctx context.Context,
 		configs = append(configs, *configConfigMap)
 	}
 
+	// Look for operator generated MachineConfigs from the hosted control plane namespace
+	operatorGeneratedConfigs := &corev1.ConfigMapList{}
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:1308]: `operatorGeneratedConfigs`, which operator?

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 8 above)
Comment (enxebre) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:75]: also can you be consistent for variable naming and value with 
tunedConfigKey      = "tuned"
tunedConfigMapLabel = "hypershift.openshift.io/tuned-config"

and group all the variable declarations together?

Comment (enxebre) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:75]: can you also add a comment to reconcileTunedConfigMap elaborating that it is just mirroring the config into the control plane namespace for the nto operator consumption (which then mirror them into the guest cluster for the daemon consumption?).

File: docs/content/how-to/node-tuning.md (hunk 2 above)
Comment (dagrayvid) [reply docs/content/how-to/node-tuning.md:4]: Thanks, I wasn't aware of this. I changed the link to latest.

File: docs/content/how-to/node-tuning.md (hunk 4 above)
Comment (dagrayvid) [reply docs/content/how-to/node-tuning.md]: done

File: docs/content/how-to/node-tuning.md (hunk 5 above)
Comment (dagrayvid) [reply docs/content/how-to/node-tuning.md]: fixed, thanks.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 6 above)
Comment (dagrayvid) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: Thanks @enxebre. As discussed on the project-hypershift slack, I rewrote this to get the HostedControlPlane from the control plane namespace, and check it's hosted cluster (by annotation) to get the namespace of the NodePool.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 7 above)
Comment (dagrayvid) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: I fixed this.  I am now using hyperv1.NodePoolLabel which is the exact label I am getting from the Nodes on the NTO side, and copying to the ConfigMap.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 8 above)
Comment (dagrayvid) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: Thanks, I think I addressed all of these comments. I made the label NTO-specific, and grouped it with the other tuned* labels, and added the comment you requested.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 9 above)
Comment (dagrayvid) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: Thanks, as discussed on slack I was originally thinking it may be useful to keep this generic, but since we currently have no other need for this functionality, I renamed this variable to be NTO specific (`nodeTuningGeneratedConfigs`)

**Comment by dagrayvid** [comment]
https://github.com/openshift/cluster-node-tuning-operator/pull/456 has now merged. From the NTO side, I believe this is ready for merge or further review.

/unhold

**Comment by dagrayvid** [comment]
@sjenning the e2e-aws failures look like flakes to me, can you confirm?

File: docs/content/how-to/node-tuning.md (hunk 3 above)
Comment (enxebre) [reply docs/content/how-to/node-tuning.md:134]: There should be no possible choice for the user to target particular nodes within a NodePool.

File: docs/content/how-to/node-tuning.md [hunk 10]
```diff
@@ -95,4 +96,126 @@ If you would like to set some node-level tuning on the nodes in your hosted clus
...
+   
+    List the Profiles in the hosted cluster:
+    ```
+    oc --kubeconfig="$HC_KUBECONFIG" get Profiles -n openshift-cluster-node-tuning-operator
+    ```
+
+    Example output:
+    ```
+    NAME                           TUNED                      APPLIED   DEGRADED   AGE
+    nodepool-1-worker-1            openshift-node             True      False      132m
```
Comment (enxebre) [inline docs/content/how-to/node-tuning.md:204]: what happens if someone modified this profiles?

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 11]
```diff
@@ -1303,6 +1308,17 @@ func (r *NodePoolReconciler) getConfig(ctx context.Context,
 		configs = append(configs, *configConfigMap)
 	}
 
+	// Look for operator generated MachineConfigs from the hosted control plane namespace
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:1311]: nit: look for nto generated...
Might seems obvious but people with less context would wonder

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 12]
```diff
@@ -1656,6 +1686,21 @@ func (r *NodePoolReconciler) enqueueNodePoolsForConfig(obj client.Object) []reco
 	return result
 }
 
+// getNodePoolNamespace returns the namespaced name of a NodePool, given the NodePools name
+// and the control plane namespace name for the hosted cluster that this NodePool is a part of.
+func (r *NodePoolReconciler) getNodePoolNamespacedName(nodePoolName string, controlPlaneNamespace string) (types.NamespacedName, error) {
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:1691]: can we please unit test this?

File: api/v1alpha1/hostedcluster_types.go [hunk 13]
```diff
@@ -114,6 +114,9 @@ const (
 	// removed when deleting the corresponding HostedCluster. If set to "true", resources created on the cloud provider during the life
 	// of the cluster will be removed, including image registry storage, ingress dns records, load balancers, and persistent storage.
 	CleanupCloudResourcesAnnotation = "hypershift.openshift.io/cleanup-cloud-resources"
+
+	// HostedClusterAnnotation is an annotation that indicates which HostedCluster an object corresponds to by namespace/name
```
Comment (enxebre) [inline api/v1alpha1/hostedcluster_types.go:118]: can you remind me why was this moved here? seems we are only consumers ourselves

**Comment by enxebre** [comment]
Can we please have e2e tests in place for both in-place / replace, similar to https://github.com/openshift/hypershift/blob/main/test/e2e/nodepool_machineconfig_test.go

File: api/v1alpha1/hostedcluster_types.go (hunk 13 above)
Comment (dagrayvid) [reply api/v1alpha1/hostedcluster_types.go:118]: I had asked whether we should move it to the API if it is being used in the NodePool controller. I can move it back to hostedcluster_controller.go and just import the hostedcluster package in nodepool_controller.go if you prefer that over adding it to the API.

File: docs/content/how-to/node-tuning.md (hunk 3 above)
Comment (dagrayvid) [reply docs/content/how-to/node-tuning.md:134]: This is behaviour is based off of how NTO works in standalone OCP. There are many cases where a TuneD profile is making in-place changes to node tunables and no MachineConfig is needed. For example, if a user wants to set some sysctl values on one node with particular labels and assign some Pods only to that Node by label. 

If we do decide to remove this feature in HyperShift, we can do that, but it would be a change to the NTO code.

Comment (dagrayvid) [reply docs/content/how-to/node-tuning.md:134]: Also note that the original issue Jiri hit here was fixed. If a user does use Node label based matching, no MachineConfig will be generated based on that Profile

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 12 above)
Comment (dagrayvid) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:1726]: Thanks @enxebre, I added a unit test for this function.

**Comment by dagrayvid** [comment]
> Can we please have e2e tests in place for both in-place / replace validating e.g. what you describe in the docs, similar to https://github.com/openshift/hypershift/blob/main/test/e2e/nodepool_machineconfig_test.go

Thanks @enxebre, I added e2e tests for in-place / replace of the workflow described in the docs.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 11 above)
Comment (dagrayvid) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: Thanks, I fixed this comment.

File: docs/content/how-to/node-tuning.md (hunk 10 above)
Comment (dagrayvid) [reply docs/content/how-to/node-tuning.md:204]: @enxebre I'm not sure I know what you are asking. Are you wondering what would happen if someone modified the Profile objects from the hosted cluster side?

**Comment by dagrayvid** [comment]
The kubevirt-e2e-kubevirt-azure-ovn failures seem unrelated to the PR. 

/retest

**Comment by dagrayvid** [comment]
@enxebre I believe all the suggestions have been implemented.  I did have a few questions about some of your comments. 

Otherwise I think this PR is nearly ready to merge (or for further review).

File: docs/content/how-to/node-tuning.md (hunk 10 above)
Comment (enxebre) [reply docs/content/how-to/node-tuning.md:204]: >Are you wondering what would happen if someone modified the Profile objects from the hosted cluster side?

Yes, is it possible to change something guest cluster side which the NTO watches and reconciles against management side config and so triggering an upgrade?

Comment (dagrayvid) [reply docs/content/how-to/node-tuning.md:204]: In theory, yes. This was discussed in some of the earlier design discussions about enabling NTO on HyperShift. Like in standalone OCP, the NTO Operand (containerized TuneD daemon) writes the needed kernel boot parameters calculated by TuneD based on the applied profile to the Profile objects status.bootcmdline field. This field is read by the Operator before creating / updating the NTO-generated MachineConfig.

If the Profile object were edited by someone with admin privileges on the guest cluster, the Operator and Operand would simultaneously reconcile. The Operand would reconcile the Profile, overwriting any change in the status. The Operator would also reconcile the Profile, potentially updating the NTO-generated MachineConfig based on the Profile status.bootcmdline change, in a race with the Operand. If the operator "loses" the race, after the operand does overwrite any admin user changes to the Profile, the operator will reconcile the Profile again, syncing the MachineConfig.

When we discussed this earlier on, the answer was that this should be "okay" as admin users of the hosted cluster already have root access to the nodes (i.e. oc debug).

**Comment by dagrayvid** [comment]
> Would NTO ever try to do anything either in management or guest side without expressing intent via nodepool.spec?

@enxebre I'm not sure I understand this question. NTO cannot express intent via the nodepool.spec as the NodePool is outside of the hosted control plane namespace, so NTO doesn't have the required RBAC privileges to modify it. For regular guest-side tuning changes (from the already merged functionality), NTO creates / applies TuneD profiles that are in the ConfigMaps generated by the NodePool controller, created in the HCP namespace.

In this PR, we are also making it so that the NTO-generated ConfigMaps containing MachineConfigs (which are created by NTO in the HCP namespace) are reconciled by the NodePool controller. These are not added to the NodePool.spec.

**Comment by enxebre** [comment]
hey @dagrayvid 

>In this PR, we are also making it so that the NTO-generated ConfigMaps containing MachineConfigs (which are created by NTO in the HCP namespace) are reconciled by the NodePool controller. These are not added to the NodePool.spec.

That's what I mean. Given the behaviour introduced by this PR:
Would it possible to trick the NTO to generate/modify a NTO-generate config that in turn would trigger a rolling upgrade by any other mechanism than modifying nodePool.spec?

If I modify the nto resources existing in guest cluster, would NTO watch that and modify a NTO-generated config and so causing a rolling upgrade?

Is this PR enabling in any way someone with access to the guest cluster to effectively trigger a rolling upgrade?

**Comment by dagrayvid** [comment]
> That's what I mean. Given the behaviour introduced by this PR: Would it possible to trick the NTO to generate/modify a NTO-generate config that in turn would trigger a rolling upgrade by any other mechanism than modifying nodePool.spec?
> 
> If I modify the nto resources existing in guest cluster, would NTO watch that and modify a NTO-generated config and so causing a rolling upgrade?
> 
> Is this PR enabling in any way someone with access to the guest cluster to effectively trigger a rolling upgrade?

Yes, I think that is theoretically possible currently. I'm not sure of a feasible alternative, because the kernel boot cmdline needs to be calculated on the guest side -- on the Node by the (containerized) TuneD daemon and communicated back to NTO and to the NodePool controller.

One not fully fleshed-out idea: the NodePool controller could NOT watch/reconcile the NTO-generated MachineConfigs, and instead require some additional admin step on the management side before triggering the rolling update? This would be a big change in user experience from standalone OCP. It would require two steps to apply tuning that requires kernel boot parameters, first create the Tuned in a ConfigMap and reference it in the NodePool, then wait a few seconds for the TuneD daemon to update the profile.status.bootcmdline and for NTO to generate the MachineConfig. Then the admin on the management side can "trigger" the rolling update based on this calculated MachineConfig. 

Maybe the trigger could be adding this NTO generated ConfigMap to the NodePool.spec.config. Can a ConfigMap in the HCP namespace be referenced in nodepool.spec.config? We would need to ignore these NTO-generated machineconfigs in [enqueueNodePoolsForConfig](https://github.com/openshift/hypershift/blob/026a43c851b03fea6ab9414591cda32f11a5c668/hypershift-operator/controllers/nodepool/nodepool_controller.go#L1663-L1688) to ensure changes on the hosted cluster side aren't automatically applied (triggering an update). This idea seems prone to user errors where the user isn't sure why their bootcmdline isn't updated when they expect it to be...
cc @jmencak

File: api/v1alpha1/hostedcluster_types.go (hunk 13 above)
Comment (dagrayvid) [reply api/v1alpha1/hostedcluster_types.go]: @enxebre, I undid this change. I don't think there is any point in moving this here, as we are already importing the hostedcluster package in nodepool_controller.go. It keeps the PR much cleaner this way.

**Comment by dagrayvid** [comment]
Thanks @enxebre. I rebased this PR on the latest changes and decided to keep `hostedcluster.HostedClusterAnnotation` as it was (out of the API) for the sake of simplicity. Tested locally.

**Comment by dagrayvid** [comment]
/retest

Failures seem unrelated to the PR. Last failure was hitting the 1h timeout without any failed tests

**Comment by enxebre** [comment]
/lgtm
/hold
to make sure @sjenning have the chance to have look

**Comment by sjenning** [comment]
lgtm
/hold cancel
e2e flaked on https://github.com/openshift/hypershift/pull/1798 before
/retest-required
